[NativeAOT-LLVM]: Struct padding on the shadow stack

yowl opened the pull request:
This PR adds a call back to calculate the padding required for structs on the shadowstack, and removes the forced failures for this scenario.  Also 

- Allows multiple returns
- Changes `OperIsBlkOp ` to explicitly check the operands instead of `Data()  != nullptr` <-  Shall I create a PR to push that upstream?
- For Blk operations sets the layout and `gtBlkOpKind`. I fixed it to  `BlkOpKindHelper`  which is just a guess.

yowl committed:
Merge remote-tracking branch 'origin/feature/NativeAOT-LLVM' into feature/NativeAOT-LLVM

yowl committed:
allow multiple return blocks

OperIsBlkOp checks oper explicitly

Allow structs on shadow stacks by adding the call to get the padding and offset

yowl committed:
tidy, remove comments and unused ifs

SingleAccretion commented on src/coreclr/jit/gentree.h:
```diff
@@ -7181,7 +7181,7 @@ struct GenTreeCC final : public GenTree
 inline bool GenTree::OperIsBlkOp()
 {
-    return ((gtOper == GT_ASG) && varTypeIsStruct(AsOp()->gtOp1)) || (OperIsBlk() && (AsBlk()->Data() != nullptr));
+    return ((gtOper == GT_ASG) && varTypeIsStruct(AsOp()->gtOp1)) || (OperIsBlk() && (OperIs(GT_STORE_DYN_BLK, GT_STORE_BLK, GT_STORE_OBJ)));
```
```suggestion
    return ((gtOper == GT_ASG) && varTypeIsStruct(AsOp()->gtOp1)) || OperIs(GT_STORE_DYN_BLK, GT_STORE_BLK, GT_STORE_OBJ);
```
Since `OperIsBlk` just checks for these same opers.

>  Shall I create a PR to push that upstream?

Not-a-maintainer's-opinion: yes.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -407,12 +406,11 @@ unsigned int Llvm::padNextOffset(CorInfoType corInfoType, unsigned int atOffset)
-        // var size = type is DefType && type.IsValueType ? ((DefType)type).InstanceFieldSize
-        //                                               : type.Context.Target.LayoutPointerSize;
-        failFunctionCompilation(); // TODO value type sizes and alignment
+        // TODO-LLvm: LCLBLK size?
```
...will always be zero as the outgoing args size is calculated during simple lowering.

I think we can just ignore it.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1185,11 +1183,11 @@ int Llvm::getLocalOffsetAtIndex(GenTreeLclVar* lclVar)
                 if (!canStoreLocalOnLlvmStack(varDsc))
                 {
-                    offset = padNextOffset(corInfoType, offset);
+                    offset = padNextOffset(corInfoType, varDsc->lvClassHnd, offset);
```
What is the reason for using `lvClassHnd`? It is only used for type optimizations in the compiler and I suspect is not available in a number of cases. I would have expected we would use `GetStructHnd` here.

yowl replied:
Right, thanks

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1466,7 +1464,13 @@ void Llvm::ConvertShadowStackLocalNode(GenTreeLclVarCommon* node)
+            GenTreeBlk* blk = node->AsBlk();
+            CORINFO_CLASS_HANDLE handle = varDsc->GetStructHnd();
+            blk->SetLayout(_compiler->typGetObjLayout(handle));
+            blk->gtBlkOpKind = GenTreeBlk::BlkOpKindHelper; // TODO-LLVM: dumping the tree requires a valid value, is this ok?
```
I would say leave this as `BlkOpKindInvalid` - we won't be using it.

Normally it is set during lowering so that the register allocator knows how many temporary registers are needed for the copy, but we'll have LLVM figure this out by itself.

yowl replied:
thanks

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1634,9 +1651,10 @@ void Llvm::PlaceAndConvertShadowStackLocals()
         LclVarDsc* varDsc = locals.at(i);
         CorInfoType corInfoType = toCorInfoType(varDsc->TypeGet());
-        offset = padOffset(corInfoType, offset);
+        CORINFO_CLASS_HANDLE classHandle = corInfoType == CORINFO_TYPE_VALUECLASS ? varDsc->GetStructHnd() : varDsc->lvClassHnd;
```
Ditto: `lvClassHnd`. Is there a specific need to pass this down?

yowl replied:
I'm changing this to be clearer I hope, it will just have the struct class handle.

SingleAccretion commented on src/coreclr/jit/llvm.h:
```diff
@@ -174,8 +176,8 @@ class Llvm
     bool needsReturnStackSlot(CorInfoType corInfoType, CORINFO_CLASS_HANDLE classHnd);
-    unsigned int padNextOffset(CorInfoType corInfoType, unsigned int atOffset);
-    unsigned int padOffset(CorInfoType corInfoType, unsigned int atOffset);
+    unsigned int padNextOffset(CorInfoType corInfoType, CORINFO_CLASS_HANDLE classHandle, unsigned int atOffset);
+    unsigned int padOffset(CorInfoType corInfoType, CORINFO_CLASS_HANDLE classHandle, unsigned int atOffset);
```
General question: do you expect we'll leave this "in the VM" in the long term or is this a temporary solution to have the IL/RyuJit - codegens agree with each other? I ask because if we think we should leave this logic to the VM, we should probably think of ways to make transmitting the results cheaper (RPI transitions in loops seem expensive).

yowl replied:
We'll need it long term I think.  The VM defines the layout I believe and we need the alignment and packing size from that.  If we dont use the same layout then I think structs in structs for example, would be a problem.

yowl committed:
Update src/coreclr/jit/gentree.h

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
remove spilled expressions

address feedback on class handle use

yowl committed:
Merge branch 'struct-padding' of https://github.com/yowl/runtimelab into struct-padding

yowl commented:
I tidied this up addressing the feedback.  I removed the `spilledExpression` vector, there was nothing in it at present.  It can be put back when/if  required.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -375,8 +378,12 @@ CorInfoType Llvm::toCorInfoType(var_types varType)
 
+CORINFO_CLASS_HANDLE Llvm::tryGetStructClassHandle(LclVarDsc* varDsc)
+{
+    return toCorInfoType(varDsc->TypeGet()) == CorInfoType::CORINFO_TYPE_VALUECLASS ? varDsc->GetStructHnd() : nullptr;
```
```suggestion
    return varTypeIsStruct(varDsc) ? varDsc->GetStructHnd() : NO_CLASS_HANDLE;
```

yowl replied:
Thanks, I'll get that in the next PR.

SingleAccretion reviewed (approved):
LGTM, modulo one nit - feel free to address separately (to not spin up the CI again).

yowl commented:
@jkotas   I'll get the last suggestion in the next PR.  Pending comments from yourself and others this is ok to go.

jkotas reviewed (approved):
@yowl @SingleAccretion It is a pleasure to see you working together on this!
